### PR TITLE
Status codes for Aggregation errors, part 2

### PR DESCRIPTION
**Pull request by not-napoleon**
Continuation of https://github.com/elastic/elasticsearch/pull/99963

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregator.java [hunk 1]
```diff
@@ -84,12 +85,14 @@ public final InternalAggregation doReduce(Aggregations aggregations, Aggregation
...
                             }
                             if (currElement == parsedPath.size() - 1) {
+                                // Seems not retryable, should be 400
                                 throw new AggregationExecutionException(
                                     "invalid bucket path ends at [" + parsedPath.get(currElement).key() + "]"
                                 );
                             }
                             currentAgg = bucket.getAggregations().get(parsedPath.get(++currElement).name());
                         } else {
+                            // Seems not retryable, should be 400
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregator.java:95]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregator.java [hunk 2]
```diff
@@ -107,6 +110,7 @@ public final InternalAggregation doReduce(Aggregations aggregations, Aggregation
                     }
                 }
                 if (currentAgg instanceof InternalMultiBucketAggregation == false) {
+                    // Seems not retryable, should be 400
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregator.java:113]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java [hunk 3]
```diff
@@ -183,6 +183,8 @@ public <T> T getAggregator(RegistryKey<T> registryKey, ValuesSourceConfig values
             }
             return supplier;
         }
+        // This should be a startup error.  Should never happen, probably indicates a bad plugin if it does.  Should probably log and have
+        // actual docs on how to resolve.
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java:187]: :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/cumulativecardinality/CumulativeCardinalityPipelineAggregator.java [hunk 4]
```diff
@@ -81,6 +81,7 @@ private AbstractHyperLogLogPlusPlus resolveBucketValue(
         List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
         Object propertyValue = bucket.getProperty(agg.getName(), aggPathsList);
         if (propertyValue == null) {
+            // Seems like a user error, change to 400
```
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/cumulativecardinality/CumulativeCardinalityPipelineAggregator.java:84]: :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/cumulativecardinality/CumulativeCardinalityPipelineAggregator.java [hunk 5]
```diff
@@ -98,6 +99,7 @@ private AbstractHyperLogLogPlusPlus resolveBucketValue(
         }
 
         throw new AggregationExecutionException(
+            // seems like a user error, change to 400
```
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/cumulativecardinality/CumulativeCardinalityPipelineAggregator.java:102]: :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/InternalMultiTerms.java [hunk 6]
```diff
@@ -194,6 +194,7 @@ static class TermsComparator implements Comparator<List<Object>> {
         @Override
         public int compare(List<Object> thisTerms, List<Object> otherTerms) {
             if (thisTerms.size() != otherTerms.size()) {
+                // Not clear on how this can happen.
```
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/InternalMultiTerms.java:197]: No idea!

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/GeoLineMultiValuesSource.java [hunk 7]
```diff
@@ -25,6 +25,7 @@ public GeoLineMultiValuesSource(Map<String, ValuesSourceConfig> valuesSourceConf
         for (Map.Entry<String, ValuesSourceConfig> entry : valuesSourceConfigs.entrySet()) {
             final ValuesSource valuesSource = entry.getValue().getValuesSource();
             if (valuesSource instanceof ValuesSource.Numeric == false && valuesSource instanceof ValuesSource.GeoPoint == false) {
+                // This will not resolve on retry, should be 400
```
Comment (nik9000) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/GeoLineMultiValuesSource.java:28]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java [hunk 8]
```diff
@@ -64,6 +64,7 @@ private static void executeInSortOrder(SearchContext context, BucketCollector co
         try {
             searcher.search(context.rewrittenQuery(), collector);
         } catch (IOException e) {
+            // Seems like this should be 400 (non-retryable), but we clearly intentionally throw a 500 here.  Why?
             throw new AggregationExecutionException("Could not perform time series aggregation", e);
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java:68]: This one is confusing to me; why do we make it a time series specific error, and not just directly bubble up the underlying IOException?

**Comment by not-napoleon** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java:68]: Right, this exception wrapping doesn't make a lot of sense. Maybe just wrap it in a `UncheckedIOException`? (The method and the method calling this method don't expect a IOException)

**Review by martijnvg — APPROVED** [review]
lgtm - if some of these programming exceptions are thrown then we need to investigate, which means it is likely a bug.

**Pull request merged**